tests/mock_crust/tunnel.rs

```diff
@@ -147,3 +147,74 @@ fn tunnel_clients() {
                                          min_section_size);
     verify_invariant_for_all_nodes(&nodes);
 }
+
+#[test]
+fn tunnel_peer_connect_failure() {
+    let min_section_size = 5;
+    let network = Network::new(min_section_size, None);
+    network.block_connection(Endpoint(2), Endpoint(3));
+    network.block_connection(Endpoint(3), Endpoint(2));
+    let mut nodes = create_connected_nodes(&network, min_section_size);
+
+    network.send_crust_event(Endpoint(2), crust::Event::ConnectFailure(PeerId(3)));
+    let _ = poll_all(&mut nodes, &mut []);
+    verify_invariant_for_all_nodes(&nodes);
+}
+
+// Verifies tunnel switched.
+fn verify_tunnel_switch(nodes: &mut Vec<TestNode>, node: usize, client_1: usize, client_2: usize) {
+    while let Ok(event) = nodes[client_1].inner.try_next_ev() {
+        match event {
+            Event::NodeLost(name, _) => {
+                assert!(name == nodes[node].name() || name == nodes[client_2].name());
```

Comment:
I guess we could tighten this up slightly by ensuring that `client_1` receives exactly two `NodeLost`s then one `NodeAdded`, and that `client_2` receives one `NodeLost` followed by one `NodeAdded`?